python/rmm/_lib/device_buffer.pyx

```diff
@@ -344,7 +344,7 @@ cdef void _copy_async(const void* src,
                       void* dst,
                       size_t count,
                       ccudart.cudaMemcpyKind kind,
-                      cuda_stream_view stream) nogil:
+                      cuda_stream_view stream) nogil except *:
```

Comment:
Nice catch, John! Should  this function even be marked `nogil`? It clearly needs the GIL to raise the `RuntimeError` that it does

Reply:
Yeah it's an interesting question. Was thinking this might be permissible, but maybe only in Cython 3 ( https://github.com/cython/cython/issues/3558 )

Reply:
Suppose there are a few options:

1. Drop the `nogil`
2. Use a `with gil` block in this function if an error comes up
3. Return `-1` and raise outside this function
4. ?

Reply:
Good call, I didn't even think about that. I think we should probably not mark the function nogil, but just wrap the `cudaMemcpyAsync` call in a `with nogil` block.

Reply:
Sounds good to me

Reply:
It's worth noting there is some overhead to releasing the GIL. So if the function is fast, it might not be worth releasing the GIL.

Reply:
I think we can leave this as `nogil` even with the `raise`?

> Oh, and the `with gil:` is no longer required for the raise since 0.29. It's injected automatically when `raise` is used in a `nogil` section.

https://github.com/cython/cython/issues/3558#issuecomment-620592470

Also confirmed in this [Changelog entry from 0.29:](https://cython.readthedocs.io/en/latest/src/changes.html#id111)

> Raising exceptions from nogil code will automatically acquire the GIL, instead of requiring an explicit with gil block.

Reply:
Thanks Bradley! 🙏

Was looking for that text in the docs as I was thinking that might be the case, but only found it in relation to 3.0.0. Missed the changelog entry.

As we already [require Cython 0.29 as a minimum]( https://github.com/rapidsai/rmm/blob/89713b2c221f8931044d1e5c7a5889cad4125454/conda/recipes/rmm/meta.yaml#L34 ), we should be good here.